cuda/distributed/vector_kernels.cu

```diff
@@ -36,22 +36,27 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <ginkgo/core/base/exception_helpers.hpp>
 
 
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/scatter.h>
+#include <thrust/tuple.h>
+
+
 namespace gko {
 namespace kernels {
 namespace cuda {
 namespace distributed_vector {
 
 
-template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
-void build_local(
-    std::shared_ptr<const DefaultExecutor> exec,
-    const device_matrix_data<ValueType, GlobalIndexType>& input,
-    const distributed::Partition<LocalIndexType, GlobalIndexType>* partition,
-    comm_index_type local_part,
-    matrix::Dense<ValueType>* local_mtx) GKO_NOT_IMPLEMENTED;
+#define GKO_THRUST_LAMBDA __device__
```

Comment:
I think we can just use __device__ directly, which is also shorter

Reply:
could you try replacing the generic lambdas by explicit ones and see if they still occur? I haven't had any issues related to this so far.

Reply:
Even with the generic lambdas I get the same error message. Btw, I'm using rocm 4.3, if that makes a difference. If you are interested, this is the error message:
```
/opt/rocm-4.3.0/include/thrust/detail/type_traits.h:455:26: error: no type named 'type' in 'thrust::detail::result_of_adaptable_function<(lambda at /home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:61:9) (thrust::tuple<int, unsigned long>)>'
  typedef typename Then::type type;
          ~~~~~~~~~~~~~~~^~~~
/opt/rocm-4.3.0/include/thrust/iterator/detail/iterator_adaptor_base.h:48:5: note: in instantiation of template class 'thrust::detail::eval_if<true, thrust::detail::result_of_adaptable_function<(lambda at /home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:61:9) (thrust::tuple<int, unsigned long>)>, thrust::detail::identity_<thrust::use_default>>' requested here
  : thrust::detail::eval_if<
    ^
/opt/rocm-4.3.0/include/thrust/iterator/detail/transform_iterator.inl:38:38: note: in instantiation of template class 'thrust::detail::ia_dflt_help<thrust::use_default, thrust::detail::result_of_adaptable_function<(lambda at /home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:61:9) (thrust::tuple<int, unsigned long>)>>' requested here
    typedef typename thrust::detail::ia_dflt_help<
                                     ^
/opt/rocm-4.3.0/include/thrust/iterator/transform_iterator.h:191:22: note: in instantiation of template class 'thrust::detail::transform_iterator_base<(lambda at /home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:61:9), thrust::zip_iterator<thrust::tuple<const int *, unsigned long *>>, thrust::use_default, thrust::use_default>' requested here
    : public detail::transform_iterator_base<AdaptableUnaryFunction, Iterator, Reference, Value>::type
                     ^
/home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:68:25: note: in instantiation of template class 'thrust::transform_iterator<(lambda at /home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:61:9), thrust::zip_iterator<thrust::tuple<const int *, unsigned long *>>>' requested here
    auto local_row_it = thrust::make_transform_iterator(
                        ^
/home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:94:5: note: in instantiation of function template specialization 'gko::kernels::hip::distributed_vector::build_local<std::complex<double>, int, int>' requested here
    GKO_DECLARE_DISTRIBUTED_VECTOR_BUILD_LOCAL);
```

Reply:
`__host__ __device__` everywhere then? :) I guess I just didn't use transform_iterator on ROCm before, contrary to what I thought

Reply:
Yes, I will use that. I didn't use it before, because `__host__ __device__` doesn't work with cuda and generic lambdas...